sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java

```diff
@@ -945,4 +951,132 @@ public static CustomHttpErrors createBigQueryClientCustomErrors() {
             + "https://cloud.google.com/service-infrastructure/docs/rate-limiting#configure");
     return builder.build();
   }
+
+  private static class BoundedExecutorService implements ExecutorService {
```

Comment:
In general this is a less-ideal way of implementing this, because it forces the unbounded executor to create threads which then just stick around blocking on the semaphore. A better technique is to buffer the Runnables as needed rather than wrapping them..

Reply:
However in this use case the semaphore approach is probably not too bad.